### PR TITLE
rtorrent module

### PR DESCRIPTION
**Pull request by mattberther**
Very early on -- but I wanted to open the pull request to 1) see if there was interest in this, and 2) identify with the community what features should be added.

At this time, my fork only shows the torrent lists. Key things Id like to get added:
- [x] Pause/Start torrents
- [x] Delete torrents
- [x] Set upload/download speeds
- [x] Add new torrents

What else?

BTW: most of the UI code was shamelessly "borrowed" from the deluge presentation. :)

### CODE REVIEW

**Comment by Hellowlol** [comment]
Thanks for the pr. I'm sure some guys would be happy to get support for rtorrent. 

While your at it you can add it to the torrent search module. 
See https://github.com/Hellowlol/HTPC-Manager/blob/master2/modules/torrentsearch.py#L173 for a example

It also needs to support uploading a link or magnet to rtorrent for the torrent module to work. Or write a wrapper that handles that.

**Comment by Hellowlol** [comment]
@mattberther Is this pr done?

**Comment by mattberther** [comment]
No - sorry. Im still working on it - but had some stuff come up that took time away from this. Im expecting to get back to it this weekend.

**Comment by mattberther** [comment]
@Hellowlol I believe this is complete - only thing Im waiting on is an icon from rg-icons. There's an open issue logged there: https://github.com/raysguy/rg-icons/issues/2

Want to make a final review and let me know if you have any changes?

**Comment by Hellowlol** [comment]
Thanks. Ill review it later today.

**Comment by Hellowlol** [comment]
I havnt tested it but it looks fine.

I just have some comments regarding style.

remove unneeded ; to terminate a line
intent is 4 spaces
remove whitespace inside a  () {} etc
remove unneed (), its not needed for conditions statements in python

**Comment by mattberther** [comment]
@Hellowlol ill update the indents...

Just to be clear: the following comments are related to the js file, yes? With the second comment, Im also assuming that you want no whitespace inside any methods. Correct?

> remove unneeded ; to terminate a line
> remove whitespace inside a () {} etc

Im not following what you're asking for with this comment -- can you provide an example of what needs to be updated and Ill fix?

> remove unneed (), its not needed for conditions statements in python

Thanks!

**Comment by Hellowlol** [comment]
No i was referring the .py file. The js file looked fine.

File: modules/rtorrent.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        'ratio': torrent[3],
+        'download_payload_rate': torrent[4],
+        'upload_payload_rate': torrent[5],
+        'eta': '-1', #TODO implement eta calculation
+        'state': 'Started' if torrent[8] == 1 else 'Paused',
+        'hash': torrent[7],
+        'total_size': torrent[6]
+      })
+
+    return { 'result': results }
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove white space inside the dict so its `return {'result': results}`

File: modules/rtorrent.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    self.logger.debug("Trying to contact rtorrent via %s" % server_url)
+    server = xmlrpclib.Server(server_url);
+    return server.system.client_version()
+
+  @cherrypy.expose()
+  @require(member_of(htpc.role_user))
+  @cherrypy.tools.json_out()
+  def start(self, torrentId=False):
+    self.logger.debug("Starting torrent %s" % (torrentId))
+    server = xmlrpclib.Server(self.stored_rpcurl());
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove unneeded ;

File: modules/rtorrent.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+      return server.d.multicall("main", "d.stop")
+
+    return server.d.stop(torrentId)
+
+  @cherrypy.expose()
+  @require(member_of(htpc.role_user))
+  @cherrypy.tools.json_out()
+  def remove(self, torrentId):
+    self.logger.debug("Removing torrent %s" % (torrentId))
+    server = xmlrpclib.Server(self.stored_rpcurl());
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove unneeded ;

File: modules/rtorrent.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+      return server.d.multicall("main", "d.start")
+
+    return server.d.start(torrentId)
+
+  @cherrypy.expose()
+  @require(member_of(htpc.role_user))
+  @cherrypy.tools.json_out()
+  def stop(self, torrentId=False):
+    self.logger.debug("Stopping torrent %s" % (torrentId))
+    server = xmlrpclib.Server(self.stored_rpcurl());
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove unneeded ;

File: modules/rtorrent.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    self.logger.debug("Removing torrent %s" % (torrentId))
+    server = xmlrpclib.Server(self.stored_rpcurl());
+
+    return server.d.erase(torrentId)
+
+  @cherrypy.expose()
+  @require(member_of(htpc.role_user))
+  @cherrypy.tools.json_out()
+  def add(self, filename=None, metainfo=None):
+    server = xmlrpclib.Server(self.stored_rpcurl());
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove ;

File: modules/rtorrent.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    if (metainfo):
```
Comment (Hellowlol) [inline modules/rtorrent.py]: this should be:

```
if metainfo:
    ....
```

File: modules/rtorrent.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+      data = base64.b64decode(metainfo)
+      self.logger.debug("Adding raw torrent: %s", filename)
```
Comment (Hellowlol) [inline modules/rtorrent.py]: self.logger.debug("Adding raw torrent: %s" % filename)

File: modules/rtorrent.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      res = server.load_raw_start(xmlrpclib.Binary(data))
+      return {'error': False } if res == 0 else {'error': True }
+
+    self.logger.debug("Adding torrent: %s", filename)
+    res = server.load_start(filename)
+
+    return {'error': False } if res == 0 else {'error': True }
```
Comment (Hellowlol) [inline modules/rtorrent.py]: `return {'error': False} if res == 0 else {'error': True}`

File: modules/rtorrent.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+  #For torrent search
+  @cherrypy.expose()
+  @require()
+  @cherrypy.tools.json_out()
+  def to_client(self, link, torrentname, **kwargs):
+    self.logger.debug("Adding torrent from torrentsearch")
+    try:
+      return self.add(link)
+    except Exception as e:
+      self.logger.debug('Failed to add %s to rTorrent %s %s'(torrentname, link, e))
```
Comment (Hellowlol) [inline modules/rtorrent.py]: Add missing %s

```
self.logger.debug('Failed to add %s to rTorrent %s %s' % (torrentname, link, e))
```

File: modules/rtorrent.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      }
+    }
+
+    return result
+
+  @cherrypy.expose()
+  @require(member_of(htpc.role_user))
+  @cherrypy.tools.json_out()
+  def set_downspeed(self, speed):
+    server = xmlrpclib.Server(self.stored_rpcurl());
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove ;

File: modules/rtorrent.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+    speed = "%sk" % speed
+    self.logger.debug('Set download speed to %s' % speed)
+    result = server.set_download_rate(speed)
+    print result
```
Comment (Hellowlol) [inline modules/rtorrent.py]: remove print

**Review by Hellowlol — COMMENTED** [review]
I added some comments so my previous comments makes sense :P

**Comment by mattberther** [comment]
Thanks! I think I got em now. :)

**Pull request merged**